### DISCUSSION TITLE
Trying to mock std::thread::spawn

### DISCUSSION THREAD

**Discussion by zephod77** [asomers/mockall#455] [Questions] [answered]
I have some code that started off looking like this:

use std::thread;
pub fn foo() -> bool {
    // do stuff
    thread::spawn(move || bar(x,y) );
    // do other stuff
   true
}
fn (x, y)
{
...
}

I want to be able to unit test foo() and mock out thread::spawn() so  I modified foo() like this
#[cfg(not(test))]
use std::thread as MyThread;
#[cfg(test)]
use MyThread as MyThread;

pub fn foo() -> bool
{
...
    MyThread::spawn::<_,u32>(move || bar(x,y) );
...
}

and in my test code I have:
mock! {
    pub JoinHandle<T> {}
}
mock! {
    pub Thread {
        pub fn spawn<F: 'static, T: 'static>(f: F) -> MockJoinHandle<T>;
   }
}

test_fn() {
    ...
    let mock_thread_ctx = MockThread::spawn_context();
    mockthread_ctx.expect::<_,u32>()
        .returning(|()| { MockJoinHandle::<u32>::new() } );
   ...
   // create test object and call foo()
}

but when I run cargo test I get 
test_fn panicked at MockThread::spawn(?) No matching expectation found.

I think this is because I don't have the parameter types correct. The u32 I added in response to the compiler complaints but I really don't care what type T is.

**Comment by asomers** [comment, marked as answer]
That's because the `F` parameter is a closure, which is impossible to name.  And that makes it impossible to create an expectation for it.  You could try this instead:
```rust
mock! {
    pub Thread {
        pub fn spawn<T: 'static>(f: Box<FnOnce() -> T> + 'static) -> JoinHandle<T>;
    }
}
```
then box your closure when you call spawn.  But if you really don't want to box closures in production, then you can work around it by defining your mocks like this:
```rust
mock! {
    pub Thread {
        pub fn _spawn<T: 'static>(f: Box<FnOnce() -> T> + 'static) -> JoinHandle<T>;
    }
}
impl MockThread {
       pub fn spawn<F, T>(f: F) -> JoinHandle
            where F: FnOnce -> T + Send + 'static,
                  T: Send + 'static
        {
            self._spawn(Box::new(f))
        }
}
```
that way you can write expectations that will take a boxed closure, but your production code won't need to do the boxing.

**Reply by zephod77** [reply]
Doesn't the signature for Thread:spawn in mock! have to exactly match the real one?

The compiler doesn't like this at all. It says "error: proc macro panicked" and so MockThread is not recognized.

**Reply by asomers** [reply]
Not quite.  Because `Thread` is a struct and not a method, the signature only needs to be close enough that your code will compile.  So if you supply a Boxed closure, either `std::Thread::spawn` or the modified `MockThread::spawn` should work.  If you rebuild Mockall with the "nightly" flag, you should get a better error message for the compile fail.  Would you show it please?
Also, I see that I made a 1-character typo in my proposal above (defining `spawn` instead of `_spawn`).  I'll fix it.

**Reply by zephod77** [reply]
OK, I'll do that. May have to be tomorrow.

**Reply by zephod77** [reply]
So I did:
$ rustup toolchain install nightly
$ cd <my_project>
$ rustup override set nightly
$ cargo test
error: proc macro panicked
  --> src/unit_tests/net_agent_unit_tests.rs:10:1
   |
10 | / mock! {
11 | |     pub Thread {
12 | |         pub fn spawn<T: 'static>(f: Box<FnOnce() -> T> + 'static) -> JoinHandle<T>;
13 | |     }
14 | | }
   | |_^
   |
   = help: message: unsupported type in this position.  More information may be available when mockall is built with the "nightly" feature.

That is exactly the same as before so I don't think I did that correctly.

**Reply by asomers** [reply]
If you use Mockall's `nightly` feature flag, the compiler will tell you exactly what's wrong.  In this case, it just wants you to add a `dyn` keyword and remove the `+ 'static`, because this is function is no longer generic on F.
```rust
mock! {
    pub Thread {
        pub fn spawn<T: 'static>(f: Box<dyn FnOnce() -> T>) -> JoinHandle<T>;
    }
}
```

**Reply by zephod77** [reply]
So it seems I can't run cargo build and cargo test. In my code I have:
MyThread::spawn::<_,()>(move || NetAgent::sub_thread(sub_skt, pub_skt) );
and cargo build works but cargo test fails and says "expected 1 generic argument.". If I remove the second generic argument then cargo build fails as says I need 2 generic arguments.

**Reply by asomers** [reply]
That's because as you've defined it the mock function has only one generic argument, but the real one has two.  However, the compiler can infer the right number.  Just leave off the turbofish: `MyThread::spawn(move ||...)`.

**Reply by zephod77** [reply]
I already tried that but the compiler said it "expected `Box<dyn FnOnce() -> _>`, found closure"
So I changed my spawn line as the compiler suggested to :
MyThread::spawn(Box::new(move || NetAgent::sub_thread(sub_skt, pub_skt) ));

which made it happy but I'm not so happy because I shouldn't have to change my code just to be able to test it.
Now the only complaint left is with regard to the expect line. I have
let mock_thread_ctx = MockThread::spawn_context();
and I want to say that I expect spawn to be called and I don't care about anything else at the moment but if I write:
mock_thread_ctx.expect_spawn::<_>()
    .returning(|_| { JoinHandle::new() });

the compiler says "no method named expect_spawn found for struct __spawn_context in the current scope. There is a method with a similar name 'expect' and of course JoinHandle doesn't have a new() method.
So I changed it to just 'expect' but what am I expecting and how do I tell it to return a JoinHandle that I'm going to ignore.

BTW for future reference, how do you use mockall's nightly flag and how do you get those code blocks? The '<>' button doesn't seem to do anything.

**Reply by asomers** [reply]
1. Yeah, it's annoying that to change one's code just for the sake of testability.  That's why I suggested the other alternative that you might want to use, the one with a mock `_spawn` method and manually implemented `spawn` method that wraps it.  It's up to you which you prefer.
2. That's correct, for static methods like this one, that have a `Context` object, the expectation method is simply `expect`.
3. [`JoinHandle`](https://doc.rust-lang.org/std/thread/struct.JoinHandle.html) doesn't provide any public constructors.  So you'll either need to do something unsafe like `unsafe{std::mem::MaybeUninit::<JoinHandle>::zeroed().assume_init()}` or else you'll actually need to create a thread in `returning` and return a real `JoinHandle`.
4. Put something like this in your `Cargo.toml` file:
```toml
mockall = { version = "0.11.3", features = [ "nightly" ] }
```
5. New to Github are you?  Do it like this: https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks#syntax-highlighting

**Reply by zephod77** [reply]
<Sigh> Well after all that I'm back to where I started. cargo test fails with "panicked at 'MockThread::spawn(?): No matching expectation found". My expect line now looks like this
```
    let mock_thread_ctx = MockThread::spawn_context(); : Context                                                                                                          
      mock_thread_ctx.expect::<u32>() ExpectationGuard<u32>                                                                                                                 
         .returning(|_: Box<(dyn FnOnce() -> _ + 'static)>| { __mockall_f:                                                                                                     
            unsafe {std::mem::MaybeUninit::<JoinHandle<_>>::zeroed().assume_init()}                                                                                           
      });            
```                         
I also tried with () instead of u32.  I can't figure out what type to expect.

**Reply by zephod77** [reply]
Ignore the ':Context' and '__mockall_f:' Thats just neovim adding annotations

**Reply by asomers** [reply]
Prop tip: If you paste something into Github and then realize you shouldn't have, you can click on the ellipses next to the comment, then click "Edit", and remove it.  As for why no matching expectation is found, are you sure that your spawned function returns a `u32`?  I'm guessing that it doesn't.  You could invoke spawn with a turbofish to be sure; if the spawned function doesn't return `u32` the compiler will complain: `MyThread::spawn(Box::new(|| 42i32)`.  If you're still stumped and want more help, you can paste your current mock structure definition and the test code too.

**Reply by zephod77** [reply]
My spawned function doesn't return anything so (). I used u32 just to see if I could get a different error.
So now my call looks like:
```
MyThread::spawn(Box::new(move || NetAgent::sub_thread(sub_skt, pub_skt));
```
and sub_thread is defined like this:
```
fn sub_thread(sub_skt: MyZmqSocket, pub_skt: MyZmqSocket)
```
The mocked thread is defined as before like this:
```
mock! {
    pub fn spawn<T: 'static>(f: Box<dyn FnOnce() -> T>) -> JoinHandle<T>;
}
```
and the expectation is set like this:
```
let mock_thread_ctx = MockThread::spawn_context();
mock_thread_ctx.expect()
    .returning(|_| { unsafe {std::mem::MaybeUninit::<JoinHandle<()>>::zeroed().assume_init()} });
```

**Reply by asomers** [reply]
Do you have any more questions?

**Reply by zephod77** [reply]
Well I'm still getting the test failure that says "panicked at 'MockThread::spawn(?): No matching expectation found".

**Reply by asomers** [reply]
The `expect` method is generic, and Rust is inferring what generic parameter you want to call it with.  Try making that explicit.  It won't fix the failure, but you'll get a much better error message.  Just explicitly provide the generic parameter to both `expect()` and `spawn()`.

**Reply by zephod77** [reply]
I had already tried that and it did not give me any more information; still the same error when running the test. It does compile OK. To be clear, I changed the spawn call to:
```
MyThread::spawn::<()>(Box::new(move || NetAgent::sub_thread(sub_skt, pub_skt));
```
and the expect call to:
```
mock_thread_ctx.expect::<()>()
    .returning(|_| { unsafe {std::mem::MaybeUninit::<JoinHandle<()>>::zeroed().assume_init()} });
```
Hmmm... I do get a couple of warnings from the compiler:
warning: the type `std::thread::JoinHandle<()>` does not permit zero-initialization()
and
"this code causes undefined behavior when executed, help: use `MaybeUninit<T>` instead, and only call `assume_init` after initialization is done"

Could that be causing the problem?

**Reply by asomers** [reply]
What do you know, it is.  Usually the "undefined behavior" warnings are merely theoretical, and I assumed that you could get away with creating an invalid JoinHandle as long as you never accessed it, but in this case it does seem to be causing the problem.  Try changing your returning method to `.returning(|_| std::thread::spawn(|| ()))`.  That will create a real JoinHandle, which fixes the problem for me.  However, it might create a thread leak in your test code.  If you can't close that leak, you may need to do some refactoring to make the product code more testable.

**Reply by zephod77** [reply]
Yes! That worked for me too. 

As for the leaked thread thread, does it actually leak? The spawn docs say that:
"If the join handle is dropped, the spawned thread will implicitly be detached. ... (It is the responsibility of the program to either eventually join threads it creates _**or**_ detach them; _**otherwise**_, a resource leak will result.)"
I read that as saying that a detached thread will be dropped when it is done and since our thread does nothing, its done immediately.
In any case each test runs in parallel so I assume that when each test ends, it cleans up any resources it has used which would include the leaked thread?

If you have any patience left with me, I'd like to get rid of the Box in the spawn call and use your other method that I couldn't get to work before.
First I don't understand how you can even use a Box instead of a closure. Is it that you can use a Boxed 'Thing' instead of a 'Thing' anywhere?
The _spawn function in the MockThread implementation is called on self but self is never passed in. I tried just adding Self to the paramaters but that made it worse.

**Reply by asomers** [reply]
If dropping a JoinHandle is sufficient to join the thread, then it looks like yes any leaking will be avoided.  Regarding the Box, the idea behind my second method is that the manually implemented `spawn` method will `Box` its generic argument, allowing the automatically generated `_spawn` method to have no generic closure arguments.  Regarding `self`, try like this:
```rust
mock! {
    pub Thread {
        pub fn _spawn<T: 'static>(f: Box<FnOnce() -> T> + 'static) -> JoinHandle<T>;
    }
}
impl MockThread {
       pub fn spawn<F, T>(f: F) -> JoinHandle
            where F: FnOnce -> T + Send + 'static,
                  T: Send + 'static
        {
            Self::_spawn(Box::new(f))
        }
}
```
But in fact, unless you intend to execute the closure in your matcher or returner, you can even eliminate it from the mock function, like this:
```rust
mock! {
    pub Thread {
        pub fn _spawn<T: 'static>() -> JoinHandle<T>;
    }
}
impl MockThread {
       pub fn spawn<F, T>(_f: F) -> JoinHandle
            where F: FnOnce -> T + Send + 'static,
                  T: Send + 'static
        {
            self._spawn()
        }
}
```

**Reply by zephod77** [reply]
I got both of those to work after fixing a couple of typos but I think I like the 1st one better as its clearer to me what is happening. I'm going to mark this thread as answered.

typos: 
JoinHandle should be JoinHandle<T>
FnOnce has to be FnOnce()
and in the second example, self._spawn() should be Self._spawn()
